File: PeakElement.java
/**
 * Time Complexity - O(logn) - Binary Search
 * Space Complexity - O(1)
 * yes this code ran on leetcode
 * Critical point was where to go on either side of mid element
 */

class Solution {
    public int findPeakElement(int[] nums) {

        int l=0, r = nums.length-1;

        while(l<r){
            int mid = (l+r)>>1;
            if(nums[mid] < nums[mid+1]){
                l = mid+1;
            }else{
                r = mid;
            }
        }
        return l;
    }
}

File: peakElement.py
#Time complexity: O(log n)
#Space complexity: O(1)

class Solution:
    def findPeakElement(self, nums) -> int:

        if len(nums) == 1:
            return 0
        low = 0
        high = len(nums) -1 

        while low <= high:

            med = low + (high - low) // 2

            if ((med == 0 and nums[med] > nums[med+1]) or (med == len(nums) - 1 and nums[med] > nums[med-1]) or (nums[med] > nums[med-1] and nums[med]> nums[med+1])):
                return med

            elif (nums[med] < nums[med+1]):
                low = med + 1
            else:
                high = med - 1

        return -1

nums = [1, 2, 3, 1]
res = Solution()
output = res.findPeakElement(nums)
print(output)